### PR TITLE
Update user registration

### PR DESCRIPTION
**Pull request by davidotte**
This PR introduces the following things:
- Simplify registration/login process by allowing retries
- Ask users some additional questions during registration.

## Allowing retries
Update registration and login functions such that they return if the login/registration was successfull and the response of the server. If the attempt was not successfull, print the response of the server (e.g. 'Weak Password') and restart the registration/login process.

## Asking additional questions
After the registration process that we had before is finished, ask the user a few more questions, which are optional. We can fine-tune the exact wording. The information entered is then sent to the server in a second request. The user is already added to the database after the first request (which we already had before). I thought this was simplest way to do it, if we send everything in one request, all the answers to the questions will have to be entered again if there was a problem at the beginning, e.g. with the password.

If you have further ideas/thoughts on this please let me know :)

### CODE REVIEW

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:41]: I think if we work on this already, it would be even better (and a considerable convenience boost) if we have this loop around the mail and the password prompt separately s.t. if I mistype my mail I get a warning about this straight away. For this: run the email validator here, too:

```python
email_info = email_validator.validate_email(
                email, test_environment=test_mode
            )
```

and also have a loop to check the passwords are equal.

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:67]: here it is fine to just have one loop

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:88]: I like the choice with retries a lot, you do not need the `lower()` here anymore though

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:94]: leave the colons (:) away, as you already have a question mark, in all three lines

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:135]: add assert to only have lower cased choices
```suggestion
        """
        assert all(c.lower() == c for c in choices), "Choices need to be lower case."
```

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [inline tabpfn_client/prompt_agent.py:131]: why have a max attempts? I think this could also go for ever, no?

**Review by SamuelGabriel — CHANGES_REQUESTED** [review]
Very cool functionality added and I do like the `_choice_with_retries`, the main thing that I would think should be improved is to have a separate loop for the email address validity, the password following our rules and the passwords matching

File: tabpfn_client/prompt_agent.py
Comment (davidotte) [reply tabpfn_client/prompt_agent.py:55]: Thanks for the feedback, I have now implemented this and it is definitely a much better user experience. :)
One thing to add: We can also have the case that the entered email already exists. This might not happen very often, but I added an extra request that is sent to the server to also check this directly when the email is entered.

File: tabpfn_client/prompt_agent.py
Comment (davidotte) [reply tabpfn_client/prompt_agent.py]: Yeah, you're right, I changed it. But not sure if it will ever make a difference :grinning:

Comment (SamuelGabriel) [reply tabpfn_client/prompt_agent.py]: nah me neither, but it is one less trajectory in the code to think about. Simpler :)

File: tabpfn_client/prompt_agent.py
Comment (SamuelGabriel) [reply tabpfn_client/prompt_agent.py:55]: yeah that is great, I wondered why you implemented email validation on the server side. good thinking!

**Review by SamuelGabriel — APPROVED** [review]
Can I ask for one more tiny thing: now that we have None as an option in the DB, could you please set empty inputs to None. Right now we would still save empty strings in the DB, if I read the code right.

Otherwise: This looks great! Thanks a lot!

I used approve as I don't think there is another review necessary after fixing the None thing above.

**Comment by davidotte** [comment]
Perfect, thanks. :)
This is already done implicitly because only the strings that are not empty are sent to the server. In the test db it works with the other values being null.

**Comment by SamuelGabriel** [comment]
ah great! this is ready to be merged without a second review

**Pull request merged**
